internal/pkg/docker/orchestrator/orchestrator.go

```diff
@@ -387,39 +439,25 @@ func (o *Orchestrator) stopTask(ctx context.Context, task Task) error {
 	}
 
 	// errCh gets one error per container
-	errCh := make(chan error)
-	for name := range task.Containers {
-		name := name
-		go func() {
-			fmt.Printf("Stopping %q\n", name)
-			if err := o.docker.Stop(ctx, o.containerID(name)); err != nil {
-				errCh <- fmt.Errorf("stop %q: %w", name, err)
-				return
-			}
-
-			// ensure that container is fully stopped before stopTask finishes blocking
-			for {
-				exists, err := o.docker.DoesContainerExist(ctx, o.containerID(name))
-				if err != nil {
-					errCh <- fmt.Errorf("polling container %q for removal: %w", name, err)
-					return
-				}
-
-				if exists {
-					select {
-					case <-time.After(1 * time.Second):
-						continue
-					case <-ctx.Done():
-						errCh <- fmt.Errorf("check container %q stopped: %w", name, ctx.Err())
-						return
-					}
-				}
+	errCh := make(chan error, len(task.Containers))
+	depGraph := buildDependencyGraph(task.Containers, ctrStateRunningOrExited)
+	err := depGraph.DownwardTraversal(ctx, func(ctx context.Context, name string) error {
+		fmt.Printf("Stopping and Removing %q\n", name)
+		if err := o.docker.Stop(ctx, o.containerID(name)); err != nil {
+			errCh <- fmt.Errorf("stop %q: %w", name, err)
+			return nil
+		}
+		if err := o.docker.Rm(ctx, o.containerID(name)); err != nil {
+			errCh <- fmt.Errorf("remove %q: %w", name, err)
+			return nil
+		}
+		fmt.Printf("Stopped and Removed %q\n", name)
+		errCh <- nil
+		return nil
+	}, ctrStateRunningOrExited, ctrStateRemoved)
```

Comment:
We skip stopping & removing a container if its `ctrStateRunningOrExited`, and we assign the container `ctrStateRemoved` after processing. 

It makes sense to me that the `state`s are `string`, this _tends to_ make the `graph` extensible: you are not limited to a binary `bool` state, but you have the choice to have an arbitrary number of self-defined states.

However, it seems to me that the extensibility is more or less an illusion. We are passing the name of the `skippedStatus` and the `requiredStatus` from the client package to the `graph` package, who handles them exactly the same way regardless what the actual passed-in status name is. There is effectively only 3 states being handled, the initial, the skipped and the required. We might as well make them a enum that are only accessible from within the `graph` package.

Alternative to private enum of 3 states, we could keep the extensible string states, but let the client host the logics that handles the transition of `state`s. In our case, for example, the logic would probably be handled within the `processVertexFn`, which would probably be of type `func (ctx context.Context, vertex V)`, something like this ⬇️ 

```go
traversal.processVertexFn = func (ctx context.Context, vertex V) {
   // docker stop, docker rm for downwardTraversal
   vertex.status = ctrStateRemoved
}
```

This comment should probably better belong to #5508 than here 😂 💭